github/commands/pull_request.py

```diff
@@ -74,59 +76,113 @@ def on_select_action(self, idx):
+        ssh_url = self.pr["head"]["repo"]["ssh_url"]
+        ref = self.pr["head"]["ref"]
+
+        if self.base_remote_url.startswith("https://"):
+            url = clone_url
+        elif self.base_remote_url.startswith("git@"):
+            url = ssh_url
+        else:
+            raise RuntimeError("Cannot recognize url {}.".format(self.base_remote_url))
+
+        self.git("fetch", url, ref)
+        self.checkout_ref(ref)
         util.view.refresh_gitsavvy_interfaces(self.window, refresh_sidebar=True)
 
-    def create_branch_for_pr(self, branch_name):
-        self.window.status_message("Fetching PR commit...")
-        self.git(
-            "fetch",
-            self.pr["head"]["repo"]["clone_url"],
-            self.pr["head"]["ref"]
-        )
+    def create_branch_for_pr(self, branch_name, checkout=False, ask_set_upstream=False):
+        if not branch_name:
+            return
 
         self.window.status_message("Creating local branch for PR...")
-        self.git(
-            "branch",
-            branch_name,
-            self.pr["head"]["sha"]
-        )
+        set_upstream = False
+
+        remotes = self.get_remotes()
+        owner = self.pr["head"]["repo"]["owner"]["login"]
+        if owner == self.base_remote.owner:
+            owner = self.base_remote_name
+            # set upstream automatically if the pr is from integrated remote
+            ask_set_upstream = False
+            set_upstream = True
+
+        clone_url = self.pr["head"]["repo"]["clone_url"]
+        ssh_url = self.pr["head"]["repo"]["ssh_url"]
+        ref = self.pr["head"]["ref"]
+
+        if ask_set_upstream:
+            set_upstream = sublime.ok_cancel_dialog(
+                "Set upstream to '{}/{}'?".format(owner, ref))
+
+        if set_upstream:
+            if owner not in remotes.keys():
+                if self.base_remote_url.startswith("https://"):
+                    url = clone_url
+                elif self.base_remote_url.startswith("git@"):
+                    url = ssh_url
+                else:
+                    raise RuntimeError("Cannot recognize url {}.".format(self.base_remote_url))
+
+                self.git("remote", "add", owner, url)
+
+            self.create_branch_from_remote_for_pr(branch_name, owner, ref, checkout)
+        else:
+            self.create_branch_from_sha_for_pr(branch_name, checkout)
+
+    def create_branch_from_remote_for_pr(self, branch_name, remote, remote_branch, checkout):
+        self.git("fetch", remote, remote_branch)
+        ref = "{}/{}".format(remote, remote_branch)
+        self.git("branch", branch_name, ref)
+        self.git("branch", "-u", ref, branch_name)
+
+        if checkout:
+            self.checkout_ref(branch_name)
+
+        util.view.refresh_gitsavvy_interfaces(self.window, refresh_sidebar=True)
+
+    def create_branch_from_sha_for_pr(self, branch_name, checkout):
+        self.git("branch", branch_name, self.pr["head"]["sha"])
```

Comment:
I think you forgot the "fetch" here.

Generally I don't understand why `create_branch_from_remote_for_pr` and `create_branch_from_sha_for_pr` are here, and especially why they differ so much.  From the code flow you would say they both should do the same, except one variant *also* sets upstream which roughly are one or two git calls. 

So they should be the same code, except:

```
self.git("remote", "add", owner, url)  # if the remote is new
self.git("branch", "-u", ref, branch_name)  # set actually new side-effect, setting the upstream
```

Can't we do 

```
self.git("fetch", url, self.pr["head"]["ref"])
self.git("branch", branch_name, self.pr["head"]["sha"])
if set_upstream:
	self.git("remote", "add", remote, url)
	self.git("branch", "-u", "{}/{}".format(remote, remote_branch), branch_name)
if checkout:
	self.checkout_ref(branch_name)
```

Reply:
In order to set up stream
```
self.git("branch", "-u", "{}/{}".format(remote, remote_branch), branch_name)
```

We need to make sure that the remote branch is fetched from the remote.

```
        self.git("fetch", remote, remote_branch)
        ref = "{}/{}".format(remote, remote_branch)
        self.git("branch", branch_name, ref)
        self.git("branch", "-u", ref, branch_name)
```
and it cannot be simply
```
self.git("fetch", url, self.pr["head"]["ref"])
``

Reply:
So there is a difference between git fetch URL ref and git fetch remote which points to URL ref.

Reply:
Yes. 

```
self.git("fetch", url, self.pr["head"]["ref"])
```
would fetch the ref to `FETCH_HEAD` instead.

Reply:
Haha, after 53 comments you introduce a new concept "FETCH_HEAD". 

> would fetch the ref to FETCH_HEAD instead

Yeah, okay, but then `git branch branch_name FETCH_HEAD`. I argue that you only need the `remote/remote_branch` for the `branch -u` call. It may not matter, I would like the least moving parts/variants as possible for readability. 

Otherwise, I think it's okay now. 

BTW. Have you seen "View diff" is broken. Github changed its server configuration and now sends lowercase headers.  Basically we look for "Location: ..." and they send "location: ...". Or "content-type"...

Reply:
> I argue that you only need the remote/remote_branch for the branch -u call

I am not sure if I understand.

> BTW. Have you seen "View diff" is broken. 

Maybe, I never used this option.